### PR TITLE
Add search by compliance and score

### PR DESCRIPTION
**Pull request by dLobatog**
Tests missing, meant to be used in the filter: https://github.com/RedHatInsights/compliance-backend/pull/103/files

### CODE REVIEW

File: app/models/concerns/system_like.rb [hunk 1]
```diff
@@ -19,4 +19,24 @@ def compliant
     end
     result
   end
+
+  def total_rules(profile = nil)
+    return profile.results(self) if profile.present?
+    profiles.map do |p|
```
Comment (akofink) [inline app/models/concerns/system_like.rb]: `profiles.flat_map`

File: app/models/concerns/system_like.rb [hunk 2]
```diff
@@ -19,4 +19,24 @@ def compliant
     end
     result
   end
+
+  def total_rules(profile = nil)
```
Comment (akofink) [inline app/models/concerns/system_like.rb]: I'm not crazy about this name. It sounds like some sort of sum. It's also returning RuleResult results for only the last import. Maybe something like `last_scan_results`?

File: app/models/concerns/system_like.rb [hunk 3]
```diff
@@ -19,4 +19,24 @@ def compliant
...
+    return profile.results(self) if profile.present?
+    profiles.map do |p|
+      p.results(self)
+    end.flatten
+  end
+
+  def rules_passed(profile = nil)
+    total_rules(profile).count { |result| result }
```
Comment (akofink) [inline app/models/concerns/system_like.rb]: Maybe we should do some caching with `@rules_passed ||= ...` to reduce SQL queries (here and anywhere else we plan on calling a method multiple times).

File: app/models/concerns/system_like.rb [hunk 4]
```diff
@@ -19,4 +19,24 @@ def compliant
...
+  end
+
+  def rules_failed(profile = nil)
+    total_rules(profile).count(&:!)
+  end
+
+  def compliance_score
+    score = (100 * (rules_passed.to_f / (rules_passed + rules_failed)))
+    score.nan? ? 0.0 : score
```
Comment (akofink) [inline app/models/concerns/system_like.rb]: If a host has no rule results, it doesn't make sense to say 0 or 100. I wonder if we should just return nil?

File: app/models/host.rb [hunk 5]
```diff
@@ -12,4 +14,23 @@ class Host < ApplicationRecord
   has_many :profiles, through: :profile_hosts, source: :profile
 
   validates :name, presence: true, uniqueness: { scope: :account_id }
+
+  class << self
+    def filter_by_compliance(key, operator, value)
```
Comment (akofink) [inline app/models/host.rb]: The operator doesn't do anything (am I using this right)?

```rb
> Host.search_for("compliant = false")
[]
> Host.search_for("compliant = true")
[#<Host:0x000055fcaf3d5660 id: "a5a31c32-0bde-4c6f-b69d-f04aaf6529b4", name: "rhel7.localdomain", created_at: Mon, 08 Apr 2019 13:14:22 UTC +00:00, updated_at: Mon, 08 Apr 2019 13:14:22 UTC +00:00, account_id: "2b83f549-8ae8-4352-84e4-b853e2428cef">,
 #<Host:0x000055fcaf3d5390 id: "1e92e15c-0227-4770-acb3-e35a54df6478", name: "rhel7-insights-client.virbr0.akofink-laptop", created_at: Mon, 08 Apr 2019 17:42:32 UTC +00:00, updated_at: Mon, 08 Apr 2019 17:42:32 UTC +00:00, account_id: "2b83f549-8ae8-4352-84e4-b853e2428cef">]
> Host.search_for("compliant != true")
[#<Host:0x000055fcaf3d5660 id: "a5a31c32-0bde-4c6f-b69d-f04aaf6529b4", name: "rhel7.localdomain", created_at: Mon, 08 Apr 2019 13:14:22 UTC +00:00, updated_at: Mon, 08 Apr 2019 13:14:22 UTC +00:00, account_id: "2b83f549-8ae8-4352-84e4-b853e2428cef">,
 #<Host:0x000055fcaf3d5390 id: "1e92e15c-0227-4770-acb3-e35a54df6478", name: "rhel7-insights-client.virbr0.akofink-laptop", created_at: Mon, 08 Apr 2019 17:42:32 UTC +00:00, updated_at: Mon, 08 Apr 2019 17:42:32 UTC +00:00, account_id: "2b83f549-8ae8-4352-84e4-b853e2428cef">]
```

File: app/models/host.rb [hunk 6]
```diff
@@ -12,4 +14,23 @@ class Host < ApplicationRecord
...
+      ids = Host.includes(:profiles).all.select do |host|
+        host.compliant.values.all?(ActiveModel::Type::Boolean.new.cast(value))
+      end.pluck(:id).map { |id| "'#{id}'" }
+      return { :conditions => "1=0" } if ids.empty?
+      { :conditions => "hosts.id IN(#{ids.join(',')})" }
+    end
+
+    def filter_by_compliance_score(key, operator, range)
```
Comment (akofink) [inline app/models/host.rb]: This is how I expected it to work:

```rb
> Host.search_for("compliance_score < 100")
> Host.search_for("compliance_score < 99 and compliance_score > 97")
```

This feels a bit awkward (am I using it right?):

```rb
> Host.search_for("compliance_score = 97-99")
```

File: app/models/schema.rb [hunk 7]
```diff
@@ -112,29 +112,14 @@
...
-      profile_results = if args['profile_id'].present?
-                          Profile.find(args['profile_id']).results(obj)
-                        else
-                          obj.profiles.map do |profile|
-                            profile.results(obj)
-                          end.flatten
-                        end
-
-      profile_results.count(&:!)
+      obj.rules_passed(Profile.find(args['profile_id']))
```
Comment (akofink) [inline app/models/schema.rb]: `s/rules_passed/rules_failed/` ?

File: app/models/concerns/system_like.rb (hunk 4 above)
Comment (dLobatog) [reply app/models/concerns/system_like.rb]: I agree, normally that situation should only happen if there was some kind of db corruption or bug.

File: app/models/host.rb (hunk 5 above)
Comment (dLobatog) [reply app/models/host.rb]: You are right, I'll change these two so the operator does something...

**Pull request merged**